### DISCUSSION TITLE
Allow prepare-osd pod to schedule on "unschedulable" nodes

### DISCUSSION THREAD

**Discussion by KKonak** [rook/rook#14714] [Q&A] [answered]
Is there a way to add a toleration to the prepare-osd pod to allow it to get scheduled on nodes that are tainted with NoSchedule? I have seen the rook operator skip over starting the job on a node that is tainted so no OSD gets created. 
If the taint cannot be tolerated is there a setting that has the operator re-scan periodically to restart this job if the number of running OSDs is below the number of expected OSDs?

**Comment by travisn** [comment]
Yes, you can add tolerations for any of the rook components. In this case, you will want to set `placement.osdprepare` to add the tolerations. See the example in the [cluster.yaml placement](https://github.com/rook/rook/blob/release-1.15/deploy/examples/cluster.yaml#L168-L191).
```yaml
spec:
  placement:
    osdprepare:
      tolerations:
      ...
```
See also the [cluster placement documentation](https://rook.io/docs/rook/latest/CRDs/Cluster/ceph-cluster-crd/#placement-configuration-settings)

**Reply by KKonak** [reply]
I have seen this option and currently have the tolerations under `all` section. So I figured if it would pick up the toleration if it worked.

For more info we are trying to tolerate the `node.kubernetes.io/unschedulable=NoSchedule` taint that gets applied from cordoning a node. Fully understand that this is not normally something you'd want to do but its a customer request... So question is if this kubernetes taint is processed differently and Rook see this and doesnt attempt to schedule. Tolerating this has worked for mon/osd/mgr/etc so far.

Ill give this a shot by explicitly calling out prepareosd section to see if that makes a difference.

**Reply by travisn** [reply, marked as answer]
Ah, Rook does skip nodes that are marked specifically as unschedulable. If Rook tries to schedule there, the node may not be ready, restarting, etc, so it seems like an odd request. But if there is a request, I'm not against making it configurable to skip that check. Rook does also check to make sure the node is online and ready. See all the checks [here](https://github.com/rook/rook/blob/58a0c834bf5bce6782de5c387f3dac732393ef33/pkg/operator/k8sutil/node.go#L56-L64). 

But if it's a temporary condition like a node drain, rook would schedule the OSDs at the next reconcile. Restarting the operator would trigger that the reconcile if needed.

**Reply by KKonak** [reply]
Thanks for the reply, I figured this type of taint was handled differently by Rook but wanted to confirm if that was the case. Perhaps a toggle to skip that check would be helpful, but not a strong need. I can look to handling this with restarting operator when node is uncordoned.

**Reply by KKonak** [reply]
Wanted to follow up on this, could we get a toggle in the helm chart to disable this check in a future release? Since we use Rook as infrastructure we dont want it to be blocked from creating, we use the cordon to kick other apps off for various reasons. Even if the node is offline/restarting it would be nice to have the prepare job exist and wait for its desired node to be ready- rather than not create the job at all and have to trigger it by restarting the operator.

**Reply by travisn** [reply]
There are a couple ways to approach this. 
1. Rook could schedule anyways if there is a node annotation: `rook.io/scheduleAlways = true`
2. Add a setting to the CephCluster CR such as: `spec.storage.scheduleAlways: true`

The first option gives granularity to control nodes individually. The second option is a blanket policy for all nodes. Do you have a preference? I'm leaning towards the 2nd option, unless there is really a need for the granularity of the first option.

**Reply by KKonak** [reply]
Thanks for the reply Travis, we dont have a need to control such granularity so the second option sounds good to me. If we have a node marked as un-schedulable we essentially want Rook to ignore it and attempt to start its pods on that node if it exists in the cluster. Do you anticipate this feature to function as a toleration for the cordon taint as well? If so we can remove that toleration on our side and allow this to encompass it.
Appreciate the consideration!

**Reply by travisn** [reply]
This feature would only skip those checks for non-schedulable or nodes that are not ready, and go ahead and schedule. Are you saying you have added those tolerations in the cephcluster CR? Then you would keep those even after this feature is enabled.

**Reply by KKonak** [reply]
Okay makes sense and I have no issue keeping the tolerations set. Just wanted to define what this toggle does exactly. 
Any rough  idea on the timeline for this feature? And for our configuration management and deployment purposes having this option in the helm chart would be a huge help to us. Thanks!

**Reply by travisn** [reply]
I just opened #14949 and if all goes well it would be included in v1.15.5 next week. The cluster helm chart can set this value as well since any value in the CephCluster CR can be set inside the `cephClusterSpec` section of the helm settings.
